protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunctionProvider.java

```diff
@@ -0,0 +1,88 @@
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.protobuf.Message;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.server.annotation.RequestConverterFunction;
+import com.linecorp.armeria.server.annotation.RequestConverterFunctionProvider;
+import com.linecorp.armeria.server.protobuf.ProtobufRequestConverterFunction.ResultType;
+
+/**
+ * Provides a {@link ProtobufRequestConverterFunction} to annotated services.
+ */
+@UnstableApi
+public final class ProtobufRequestConverterFunctionProvider implements RequestConverterFunctionProvider {
+
+    @Override
+    public RequestConverterFunction createRequestConverterFunction(Type requestType,
+                                                                   RequestConverterFunction requestConverter) {
+        final ResultType resultType = toResultType(requestType);
+        if (resultType != ResultType.UNKNOWN) {
+            return new ProtobufRequestConverterFunction(resultType);
+        } else {
+            return null;
+        }
+    }
+
+    static ResultType toResultType(Type type) {
+        if (type instanceof Class) {
+            if (isProtobufMessage((Class<?>) type)) {
+                return ResultType.PROTOBUF;
+            }
+        }
+
+        if (type instanceof ParameterizedType) {
+            final ParameterizedType parameterizedType = (ParameterizedType) type;
+            final Class<?> rawType = (Class<?>) parameterizedType.getRawType();
+            if (List.class.isAssignableFrom(rawType)) {
+                final Class<?> typeArgument = (Class<?>) parameterizedType.getActualTypeArguments()[0];
+                if (isProtobufMessage(typeArgument)) {
+                    return ResultType.LIST_PROTOBUF;
+                }
+            } else if (Set.class.isAssignableFrom(rawType)) {
+                final Class<?> typeArgument = (Class<?>) parameterizedType.getActualTypeArguments()[0];
+                if (isProtobufMessage(typeArgument)) {
+                    return ResultType.SET_PROTOBUF;
+                }
+            } else if (Map.class.isAssignableFrom(rawType)) {
+                final Type[] typeArguments = parameterizedType.getActualTypeArguments();
+                final Class<?> keyType = (Class<?>) typeArguments[0];
+                if (!String.class.isAssignableFrom(keyType)) {
+                    throw new IllegalStateException(
+                            keyType + " cannot be used for the key type of Map. " +
+                            "(expected: Map<String, ?>)");
+                }
+                if (isProtobufMessage((Class<?>) typeArguments[1])) {
+                    return ResultType.MAP_PROTOBUF;
+                }
+            }
+        }
+        return ResultType.UNKNOWN;
+    }
+
+    private static boolean isProtobufMessage(Class<?> clazz) {
+        return Message.class.isAssignableFrom(clazz);
```

Comment:
`c` class extends `MessageLite`. When build a message from protobuf, `Message` class is required to get `Builder` class. However, when serialize a message to protobuf,  `MessageLite` is enough.
However, this can confuse users. Let me just use `Message` instead of `MessageLite`.

Reply:
`MessageLite` is very rare in servers btw (though I remember an issue filed by someone that was actually using it don't remember the reasoning)

Reply:
>  I remember an issue filed by someone that was actually using it don't remember the reasoning

Maybe #3085? He said it was caused by wrong dependencies.